File: src/lib/tracing.js
import { trace, context, SpanStatusCode } from '@opentelemetry/api'

// Lazy: defer getTracer until first use so the host app can register its TracerProvider first.
let tracer
const getTracer = () => (tracer ??= trace.getTracer('mohdel'))

export const startSpan = (name, attributes, parentSpan) => {
  const parentCtx = parentSpan
    ? trace.setSpan(context.active(), parentSpan)
    : undefined
  return getTracer().startSpan(name, { attributes }, parentCtx)
}

export const endSpanOk = (span, attributes) => {
  if (attributes) span.setAttributes(attributes)
  span.setStatus({ code: SpanStatusCode.OK })
  span.end()
}

export const endSpanError = (span, err) => {
  span.recordException(err)
  span.setStatus({ code: SpanStatusCode.ERROR, message: err.message })
  span.end()
}

// W3C Trace Context — https://www.w3.org/TR/trace-context/#traceparent-header
// Format: 00-<traceId>-<spanId>-<traceFlags> (lowercase hex). Used by the unix-socket
// sidecar so trace context can cross the process boundary as a header instead of trying
// to serialize a live span object (which has circular OTel internals).
const TRACEPARENT_RE = /^00-([0-9a-f]{32})-([0-9a-f]{16})-([0-9a-f]{2})$/
const INVALID_TRACE_ID = '00000000000000000000000000000000'
const INVALID_SPAN_ID = '0000000000000000'

/**
 * Parse a W3C `traceparent` header into an OTel SpanContext.
 * Returns null if the header is missing, malformed, or contains invalid all-zero IDs.
 */
export const parseTraceparent = (header) => {
  if (!header || typeof header !== 'string') return null
  const m = header.match(TRACEPARENT_RE)
  if (!m) return null
  const [, traceId, spanId, flags] = m
  if (traceId === INVALID_TRACE_ID) return null
  if (spanId === INVALID_SPAN_ID) return null
  return {
    traceId,
    spanId,
    traceFlags: parseInt(flags, 16),
    isRemote: true
  }
}

/**
 * Build a non-recording span from a `traceparent` header so the caller can use it
 * as a `parentSpan` in `startSpan(name, attrs, parentSpan)`. Returns null if the
 * header is missing/invalid.
 */
export const remoteParentFromTraceparent = (header) => {
  const spanContext = parseTraceparent(header)
  if (!spanContext) return null
  return trace.wrapSpanContext(spanContext)
}

File: rust/thin-gate/src/tracing.rs
//! OTel trace helpers — parse W3C `traceparent` strings and build a
//! `Context` rooted at the remote parent span. Thin-gate itself does
//! not register a `TracerProvider`; it relies on the embedder or the
//! host process to set one globally before the first request. When
//! no provider is registered, `global::tracer` returns a no-op, spans
//! cost near-zero, and these helpers degrade cleanly.

use opentelemetry::trace::{
    SpanContext, SpanId, TraceContextExt, TraceFlags, TraceId, TraceState,
};
use opentelemetry::Context;

/// Parse a W3C `traceparent` string into a remote `SpanContext`.
pub fn parse_traceparent(tp: &str) -> Option<SpanContext> {
    let parts: Vec<&str> = tp.split('-').collect();
    if parts.len() != 4 || parts[0] != "00" {
        return None;
    }
    let trace_id = TraceId::from_hex(parts[1]).ok()?;
    let span_id = SpanId::from_hex(parts[2]).ok()?;
    if trace_id == TraceId::INVALID || span_id == SpanId::INVALID {
        return None;
    }
    let flags = u8::from_str_radix(parts[3], 16).ok()?;
    Some(SpanContext::new(
        trace_id,
        span_id,
        TraceFlags::new(flags),
        true, // is_remote
        TraceState::default(),
    ))
}

/// Build a `Context` rooted at the `traceparent` string, or fall back
/// to the current active context when the input is absent / malformed.
pub fn parent_context(traceparent: Option<&str>) -> Context {
    match traceparent.and_then(parse_traceparent) {
        Some(sc) => Context::current().with_remote_span_context(sc),
        None => Context::current(),
    }
}
